src/dbnode/storage/limits/query_limits.go

```diff
@@ -0,0 +1,243 @@
+			recentPeak: scope.Gauge(fmt.Sprintf("recent-peak-%s", name)),
+			total:      scope.Counter(fmt.Sprintf("total-%s", name)),
+			exceeded:   scope.Tagged(map[string]string{"limit": name}).Counter("limit-exceeded"),
+		},
+		recent: atomic.NewInt64(0),
+		stopCh: make(chan struct{}),
+	}
+}
+
+func (q *queryLimits) DocsLimit() LookbackLimit {
+	return q.docsLimit
+}
+
+func (q *queryLimits) BytesReadLimit() LookbackLimit {
+	return q.bytesReadLimit
+}
+
+func (q *queryLimits) Start() {
+	q.docsLimit.Start()
+	q.bytesReadLimit.Start()
+}
+
+func (q *queryLimits) Stop() {
+	q.docsLimit.Stop()
+	q.bytesReadLimit.Stop()
+}
+
+func (q *queryLimits) AnyExceeded() error {
+	if err := q.docsLimit.Exceeded(); err != nil {
+		return err
+	}
+	if err := q.bytesReadLimit.Exceeded(); err != nil {
+		return err
+	}
+	return nil
+}
+
+// Inc increments the current value and returns an error if above the limit.
+func (q *lookbackLimit) Inc(new int) error {
+	if q == nil {
+		return nil
+	}
+	if new < 0 {
+		return fmt.Errorf("invalid negative query limit inc %d", new)
+	}
+	if new == 0 {
+		return q.Exceeded()
+	}
+
+	// Add the new stats to the global state.
+	newI64 := int64(new)
+	recent := q.recent.Add(newI64)
+
+	// Update metrics.
+	q.metrics.recent.Update(float64(recent))
+	q.metrics.total.Inc(newI64)
+
+	// Enforce limit (if specified).
+	return q.exceeded(recent)
+}
+
+func (q *lookbackLimit) Exceeded() error {
+	return q.exceeded(q.recent.Load())
+}
+
+func (q *lookbackLimit) exceeded(recent int64) error {
+	if q.options.Limit > 0 && recent > q.options.Limit {
+		q.metrics.exceeded.Inc(1)
+		return fmt.Errorf(
+			"query aborted due to limit: name=%s, limit=%d, current=%d, within=%s",
+			q.name, q.options.Limit, recent, q.options.Lookback)
+	}
+	return nil
+}
+
+// Start initializes background processing for enforcing the query limit.
+func (q *lookbackLimit) Start() {
+	if q == nil {
+		return
+	}
+	ticker := time.NewTicker(q.options.Lookback)
+	go func() {
+		for {
+			select {
+			case <-ticker.C:
+				q.reset()
+			case <-q.stopCh:
+				ticker.Stop()
+				return
+			}
+		}
+	}()
+}
+
+func (q *lookbackLimit) Stop() {
+	if q == nil {
+		return
+	}
+	close(q.stopCh)
+}
+
+// Validate returns an error if the query stats options are invalid.
+func (opts LookbackLimitOptions) Validate() error {
+	if opts.Limit < 0 {
+		return fmt.Errorf("query limit requires limit >= 0 (%d)", opts.Limit)
+	}
+	if opts.Lookback <= 0 {
+		return fmt.Errorf("query limit requires lookback > 0 (%d)", opts.Lookback)
+	}
+	return nil
+}
+
+func (q *lookbackLimit) current() int64 {
+	return q.recent.Load()
+}
+
+func (q *lookbackLimit) reset() {
+	// Update peak gauge only on resets so it only tracks
+	// the peak values for each lookback period.
+	recent := q.recent.Load()
+	q.metrics.recentPeak.Update(float64(recent))
+
+	// Update the standard recent gauge to reflect drop back to zero.
+	q.metrics.recent.Update(0)
```

Comment:
Is this necessary? It'll get updated to the first relevant value after reset anyway

Reply:
I think this accurately reflects the val it is (and also is reflective of the existing behavior which I think we wanna preserve to compare against the new peak version)